ddtrace/llmobs/_integrations/langchain.py

```diff
@@ -56,9 +61,39 @@
 SUPPORTED_OPERATIONS = ["llm", "chat", "chain", "embedding", "retrieval", "tool"]
 
 
+def _extract_bound(instance):
+    if hasattr(instance, "bound"):
+        return instance.bound
+    return instance
+
+
 class LangChainIntegration(BaseLLMIntegration):
     _integration_name = "langchain"
 
+    _chain_steps = set()  # instance_id
+    _spans = {}  # instance_id --> span
+    _instances = WeakKeyDictionary()  # spans --> instances
+
+    def record_steps(self, instance, span):
+        if not self.llmobs_enabled:
+            return
+
+        steps = getattr(instance, "steps", [])
+        for step in steps:
+            step = _extract_bound(step)
+            self._chain_steps.add(id(step))
+
+        self.record_instance(instance, span)
+
+    def record_instance(self, instance, span):
+        if not self.llmobs_enabled:
+            return
+
+        instance = _extract_bound(instance)
+
+        self._instances[span] = instance
```

Comment:
bit funky here, but this is associating spans with their instances. while this could be done directly in `_llmobs_set_tags` for individual spans, the span linking logic works by looking for the caller and identifying if it's a chain. To do that, we need a way to associate a different span with an isntance.

This is all sorta predicated on the fact that we don't have an intermediary for associating traced steps. Unlike the `pregel_loop_tick` which we could utilize for `langgraph`, all steps of a chain are iterated over directly in `chain.invoke`, with no means for us to jump between each step as they are being executed, only before and making these kinds of connections and setting these variables.

Reply:
Is there any reason we need to index by span objects rather than a simple spanID key?

Reply:
oh yeah good point, will try that out!